File: backend/amm.py
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any

BPS = 10_000
FEE_BPS = 100


@dataclass(frozen=True)
class PoolState:
    sat: int
    usd: int
    shares: int
    events: tuple[dict[str, Any], ...] = ()

    def validate(self) -> None:
        values = (self.sat, self.usd, self.shares)
        if any(value < 0 for value in values):
            raise ValueError("pool reserves and shares cannot be negative")
        if any(value == 0 for value in values) and any(value != 0 for value in values):
            raise ValueError("pool must be either empty or fully initialized")
        if any(not isinstance(event, dict) for event in self.events):
            raise ValueError("pool events must be objects")


def quote_exact_in(
    reserve_in: int, reserve_out: int, amount_in: int
) -> tuple[int, int, int]:
    if min(reserve_in, reserve_out, amount_in) <= 0:
        raise ValueError("swap amounts must be positive")
    effective = amount_in * (BPS - FEE_BPS)
    amount_out = effective * reserve_out // (reserve_in * BPS + effective)
    if amount_out <= 0 or amount_out >= reserve_out:
        raise ValueError("trade is too small or exceeds available liquidity")
    return amount_out, reserve_in + amount_in, reserve_out - amount_out


def shares_for_deposit(state: PoolState, deposit_sat: int, deposit_usd: int) -> int:
    if min(deposit_sat, deposit_usd) <= 0:
        raise ValueError("deposit amounts must be positive")
    if state.shares == 0:
        shares = math.isqrt(deposit_sat * deposit_usd)
        if shares <= 0:
            raise ValueError("initial deposit is too small to mint shares")
        return shares
    shares = min(
        deposit_sat * state.shares // state.sat,
        deposit_usd * state.shares // state.usd,
    )
    if shares <= 0:
        raise ValueError("deposit is too small to mint shares")
    return shares


def redemption_amounts(state: PoolState, shares: int) -> tuple[int, int]:
    if shares <= 0 or shares > state.shares:
        raise ValueError("shares exceed total supply")
    sat = state.sat * shares // state.shares
    usd = state.usd * shares // state.shares
    if min(sat, usd) <= 0:
        raise ValueError("LP amount is too small to redeem both assets")
    return sat, usd

File: amm.js
export const BPS = 10_000n;
export const FEE_BPS = 100n;

function requireNonNegative(value, label) {
  if (typeof value !== "bigint" || value < 0n) {
    throw new TypeError(`${label} must be a non-negative bigint`);
  }
}

function requirePositive(value, label) {
  requireNonNegative(value, label);
  if (value === 0n) throw new RangeError(`${label} must be greater than zero`);
}

export function integerSqrt(value) {
  requireNonNegative(value, "value");
  if (value < 2n) return value;

  let left = 1n;
  let right = value / 2n + 1n;
  while (left <= right) {
    const middle = (left + right) / 2n;
    const square = middle * middle;
    if (square === value) return middle;
    if (square < value) left = middle + 1n;
    else right = middle - 1n;
  }
  return right;
}

export function initialShareSupply(reserveSat, reserveUsd) {
  requirePositive(reserveSat, "reserveSat");
  requirePositive(reserveUsd, "reserveUsd");
  return integerSqrt(reserveSat * reserveUsd);
}

export function quoteExactIn(reserveIn, reserveOut, amountIn, feeBps = FEE_BPS) {
  requirePositive(reserveIn, "reserveIn");
  requirePositive(reserveOut, "reserveOut");
  requirePositive(amountIn, "amountIn");
  requireNonNegative(feeBps, "feeBps");
  if (feeBps >= BPS) throw new RangeError("feeBps must be lower than 10000");

  const amountInWithFee = amountIn * (BPS - feeBps);
  const numerator = amountInWithFee * reserveOut;
  const denominator = reserveIn * BPS + amountInWithFee;
  const amountOut = numerator / denominator;
  if (amountOut <= 0n || amountOut >= reserveOut) {
    throw new RangeError("trade is too small or exceeds available liquidity");
  }

  return {
    amountOut,
    feeAmount: amountIn - amountInWithFee / BPS,
    nextReserveIn: reserveIn + amountIn,
    nextReserveOut: reserveOut - amountOut
  };
}

export function sharesForDeposit(
  reserveSat,
  reserveUsd,
  totalShares,
  depositSat,
  depositUsd
) {
  for (const [value, label] of [
    [reserveSat, "reserveSat"],
    [reserveUsd, "reserveUsd"],
    [totalShares, "totalShares"],
    [depositSat, "depositSat"],
    [depositUsd, "depositUsd"]
  ]) requirePositive(value, label);

  const bySat = depositSat * totalShares / reserveSat;
  const byUsd = depositUsd * totalShares / reserveUsd;
  const shares = bySat < byUsd ? bySat : byUsd;
  if (shares <= 0n) throw new RangeError("deposit is too small to mint a share");
  return shares;
}

export function amountsForRedemption(
  reserveSat,
  reserveUsd,
  totalShares,
  shares
) {
  for (const [value, label] of [
    [reserveSat, "reserveSat"],
    [reserveUsd, "reserveUsd"],
    [totalShares, "totalShares"],
    [shares, "shares"]
  ]) requirePositive(value, label);
  if (shares > totalShares) throw new RangeError("shares exceed total supply");

  const sat = reserveSat * shares / totalShares;
  const usd = reserveUsd * shares / totalShares;
  if (sat <= 0n || usd <= 0n) {
    throw new RangeError("LP amount is too small to redeem both assets");
  }
  return { sat, usd };
}

export function poolPriceUsdPerBtc(reserveSat, reserveUsd) {
  requirePositive(reserveSat, "reserveSat");
  requirePositive(reserveUsd, "reserveUsd");
  return Number(reserveUsd) * 1_000_000 / Number(reserveSat);
}
